Prepare playground apps to allow testing

radical opened the pull request:
This PR has changes to prepare the `playground/` apps to be run as part of future tests in `Aspire.Playground.Tests`. The changes follow these broad patterns:

- remove all the per-project `Directory.Build.{props,targets}`, and instead of these ..
    - .. add root `Directory.Build.{props,targets}` and use `Aspire.RepoTesting.*`
    - And update all the playground projects to use `AspireProjectOrPackageReference` instead of `ProjectReference`
- `Aspire.Hosting.Testing` expects to identify AppHost assemblies by a name ending in `.AppHost`.
    - rename some projects that don't follow that convention.

- Update the apps to use `Aspire.RepoTesting*` which allows building outside of repo.

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5080)

radical committed:
[playground] Add top level Directory.Build.{props,targets}

And use Aspire.RepoTesting.* to allow testing and this replaces the
per-project sdk imports.

radical committed:
[playground] remove all the per-project Directory.Build.{props,targets}

radical committed:
[playground] Rename AppHost projects to be named consistently

.. with `.AppHost` suffix. `Aspire.Hosting.Testing` specifically looks
for a `.AppHost` suffix to identify apphost assemblies.

radical committed:
[playground] Update all ProjectReferences to AspireProjectOrPackageReference. And skip Dashboard.csproj when building for testing

radical commented:
TODO
- [ ] check that the playground apps are still runnable from VS

radical committed:
Update Aspire.sln to track the project name changes

radical committed:
Fix up import conditions

radical committed:
Merge remote-tracking branch 'origin/main' into playground-tests-prep

radical committed:
Merge branch 'main' into playground-tests-prep

eerhardt commented on playground/AWS/AWS.AppHost/AWS.AppHost.csproj:
```diff
@@ -13,10 +13,10 @@
-    <ProjectReference Include="..\..\..\src\Aspire.Hosting.AWS\Aspire.Hosting.AWS.csproj" IsAspireProjectResource="false" />    
-    <ProjectReference Include="..\..\..\src\Aspire.Hosting.AppHost\Aspire.Hosting.AppHost.csproj" IsAspireProjectResource="false" />
-    
+    <ProjectReference Include="..\..\..\src\Aspire.Dashboard\Aspire.Dashboard.csproj" Condition="'$(BuildForTest)' != 'true'" />
```
Do you think we could automatically inject this ProjectReference in a common .targets file? That way each playground project doesn't have to.

eerhardt commented on playground/AWS/AWS.AppHost/AWS.AppHost.csproj:
```diff
@@ -13,10 +13,10 @@
-    <ProjectReference Include="..\..\..\src\Aspire.Hosting.AppHost\Aspire.Hosting.AppHost.csproj" IsAspireProjectResource="false" />
-    
+    <ProjectReference Include="..\..\..\src\Aspire.Dashboard\Aspire.Dashboard.csproj" Condition="'$(BuildForTest)' != 'true'" />
+    <AspireProjectOrPackageReference Include="Aspire.Hosting.AWS" IsAspireProjectResource="false" />
```
Similar comment here for `IsAspireProjectResource="false"`. We should only need this when it is a `ProjectReference`. So when we create the `AspireProjectOrPackageReference` => `ProjectReference`, we should be able to inject this metadata.

eerhardt commented on playground/mongo/Mongo.AppHost/Mongo.AppHost.csproj:
```diff
@@ -13,12 +13,15 @@
-        
+    <ProjectReference Include="..\..\..\src\Aspire.Dashboard\Aspire.Dashboard.csproj" Condition="'$(BuildForTest)' != 'true'" />
+
+    <!-- <AspireProjectOrPackageReference Include="Aspire.Hosting" IsAspireProjectResource="false" /> -->
```
Do we need this?

eerhardt commented on playground/Directory.Build.props:
```diff
@@ -9,6 +13,8 @@
     -->
     <NoWarn>$(NoWarn),1573,1591,1712</NoWarn>
+
+    <SharedDir Condition="'$(SharedDir)' == ''">$(MSBuildThisFileDirectory)..\src\Shared\</SharedDir>
```
Why isn't this already set from the root Directory.Build.props?

radical replied:
This is to cover the case when `playground` is being built from outside the repo, and thus doesn't have that root Directory.Build.props .
I'm not super happy with this, and some others we'll need for playground, but I don't overdo that at least just yet!

eerhardt replied:
Is it needed for more than KnownResourceNames.cs?

Looks like that only contains a single const string

https://github.com/dotnet/aspire/blob/11e90df77649e3abeab69b37912b6fd7f449f50a/src/Shared/KnownResourceNames.cs#L8

Maybe we just copy that string into a `playground/KnownResourceNames.cs` file. It's just a simple string. Easier to copy it once than to try to copy around files at build time.

radical replied:
btw, there will be a bunch of copying around done on CI when archiving the playground tests. But this one is clearly a one-off, so I pushed the change as suggested.

eerhardt reviewed (commented):
This looks awesome. I had a couple suggestions to make it even cleaner. Then I think we can merge it.

radical committed:
Merge remote-tracking branch 'origin/main' into playground-tests-prep

radical committed:
Address review feedback - add Aspire.Dashboard ProjectReference to

.. `playground` `AppHost` projects by default, instead of doing that
individually in the project files.

radical committed:
Address review feedback - add IsAspireProjectResource=false

.. automatically on `@(AspireProjectOrPackageReference)` items used in
`AppHost` projects.

radical committed:
fixed missed references in playground

radical committed:
Address review feedback from @ eerhardt

Instead of copying `src/Shared` to the test archive, keep a copy of
`src/Shared/KnownResourceNames.cs` in `playground` and use that instead.

eerhardt reviewed (approved):
Nice!
